contracts/src/Gateway.sol

```diff
@@ -394,6 +416,49 @@ contract Gateway is IGateway, IInitializable {
         emit PricingParametersChanged();
     }
 
+    // @dev Register a new fungible Polkadot token for an agent
+    function _registerForeignToken(bytes32 agentID, address agent, bytes memory params) internal {
+        (bytes32 tokenID, string memory name, string memory symbol, uint8 decimals) =
+            abi.decode(params, (bytes32, string, string, uint8));
+        AssetsStorage.Layout storage $ = AssetsStorage.layout();
+        if ($.tokenRegistryByID[tokenID].isRegistered == true) {
+            revert TokenAlreadyRegistered();
+        }
+        ERC20 foreignToken = new ERC20(name, symbol, decimals);
+        address token = address(foreignToken);
+        TokenInfo memory info =
+            TokenInfo({isRegistered: true, isForeign: true, tokenID: tokenID, agentID: agentID, token: token});
+        $.tokenRegistry[token] = info;
+        $.tokenRegistryByID[tokenID] = info;
+        foreignToken.transferOwnership(agent);
+        emit ForeignTokenRegistered(tokenID, agentID, token);
+    }
+
+    // @dev Mint foreign token from Polkadot
+    function _mintForeignToken(address agent, bytes memory params) internal {
```

Comment:
The handler should be moved to Assets.sol to reduce contract size limit

Reply:
https://github.com/Snowfork/snowbridge/pull/1155/commits/3da6be486e3c02c96e77583cc3279f7dac270481